internal/config/nsmgr/nsmgr.go

```diff
@@ -0,0 +1,157 @@
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	nspkg "github.com/containernetworking/plugins/pkg/ns"
+	"github.com/containers/storage/pkg/idtools"
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+)
+
+// NamespaceManager manages the server's namespaces.
+// Specifically, it is an interface for how the server is creating namespaces.
+// and can be requested to create namespaces for a pod.
+type NamespaceManager interface {
+	// NewPodNamespaces creates new namespaces for a pod.
+	NewPodNamespaces(managedNamespaces []NSType, idMappings *idtools.IDMappings, sysctls map[string]string) ([]Namespace, error)
+}
+
+// New creates a new NamespaceManager.
+func New(namespacesDir, pinnsPath string) NamespaceManager {
+	return &namespaceManager{
+		namespacesDir: namespacesDir,
+		pinnsPath:     pinnsPath,
+	}
+}
+
+// namespaceManager is the internal implementation for NamespaceManager.
+type namespaceManager struct {
+	namespacesDir string
+	pinnsPath     string
+}
+
+// NewPodNamespaces creates new namespaces for a pod.
+// It's responsible for running pinns and creating the Namespace objects.
+// The caller is responsible for cleaning up the namespaces by calling Namespace.Remove().
+func (mgr *namespaceManager) NewPodNamespaces(managedNamespaces []NSType, idMappings *idtools.IDMappings, sysctls map[string]string) ([]Namespace, error) {
+	if len(managedNamespaces) == 0 {
+		return []Namespace{}, nil
+	}
+
+	typeToArg := map[NSType]string{
+		IPCNS:  "-i",
+		UTSNS:  "-u",
+		USERNS: "-U",
+		NETNS:  "-n",
+	}
+
+	pinnedNamespace := uuid.New().String()
+	pinnsArgs := []string{
+		"-d", mgr.namespacesDir,
+		"-f", pinnedNamespace,
+	}
+
+	if len(sysctls) != 0 {
+		pinnsArgs = append(pinnsArgs, "-s", getSysctlForPinns(sysctls))
+	}
+
+	type namespaceInfo struct {
+		path   string
+		nsType NSType
+	}
+
+	mountedNamespaces := make([]namespaceInfo, 0, len(managedNamespaces))
+
+	var rootPair idtools.IDPair
+	if idMappings != nil {
+		rootPair = idMappings.RootPair()
+	}
+
+	for _, nsType := range managedNamespaces {
+		arg, ok := typeToArg[nsType]
+		if !ok {
+			return nil, errors.Errorf("Invalid namespace type: %s", nsType)
+		}
+		pinnsArgs = append(pinnsArgs, arg)
+		pinPath := filepath.Join(mgr.namespacesDir, string(nsType)+"ns", pinnedNamespace)
+		mountedNamespaces = append(mountedNamespaces, namespaceInfo{
+			path:   pinPath,
+			nsType: nsType,
+		})
+		if idMappings != nil {
+			err := os.MkdirAll(filepath.Dir(pinPath), 0o755)
+			if err != nil {
+				return nil, err
+			}
+			f, err := os.Create(pinPath)
+			if err != nil {
+				return nil, err
+			}
+			f.Close()
+			if err := os.Chown(pinPath, rootPair.UID, rootPair.GID); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	if idMappings != nil {
+		pinnsArgs = append(pinnsArgs,
+			fmt.Sprintf("--uid-mapping=%s", getMappingsForPinns(idMappings.UIDs())),
+			fmt.Sprintf("--gid-mapping=%s", getMappingsForPinns(idMappings.GIDs())))
```

Comment:
I'd use simple string concatenation, i.e.
```go
"--uid-mapping=" + getMappingsForPinns(idMappings.UIDs())
```

Reply:
Of course, if you just move the code around it's better to make such changes in a separate commit.

Reply:
addressed